(Towards #1338) Infrastructure for Task transformation

LonelyCat124 opened the pull request:
This PR has the current implementation of the Task transformation.

Testing is still very early,m but it roughly does what is expected (and discussed somewhat in #1338).

Currently it misses doing any code gen test on a Serial region to ensure that the code is safe and valid OpenMP code as well, and that is something that needs to be decided. I may have each OMPTaskDirective store its own list of variables etc. so they can be compared in the future.

I don't think I should formally request a reviewer at this point, but if anyone does have time to have a look through the changes and see if I've done anything terrible or very wrong it would be good.

arporter commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -562,6 +566,635 @@ def _encloses_omp_directive(self):
+                         parallel_private)
+        if index_private:
+            if index_name in private_list:
+                # FIXME I think this should be
```
Just in case you're not aware (and it helps), the dependency analysis now makes use of SymPy.

arporter commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -562,6 +566,635 @@ def _encloses_omp_directive(self):
+                           .format(writer(stop_val)
+                               ,writer(start_val)))
+                else:
+                    index_name = index_name + ("-({0} - {1})"
```
I suspect you should really be building a PSyIR expression here and at line 853.

arporter commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -562,6 +566,635 @@ def _encloses_omp_directive(self):
+        out_list = []
+
+        # Create a FortranWriter to convert nodes to strings
+        # TODO: Create issue - support non-Fortran backends
```
Yes, I think if you're having to use a FortranWriter to implement functionality then there's something not quite right.

arporter commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -562,6 +566,635 @@ def _encloses_omp_directive(self):
+
+        # Find our loop initialisation, variable and bounds
+        loop_var = node.variable
+        start_val = node.children[0]
```
Semantic navigation is nicer, e.g. `start_val = node.start_expr`, `stop_expr`, `step_expr`

arporter commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -562,6 +566,635 @@ def _encloses_omp_directive(self):
+        stop_val_refs = stop_val.walk(Reference)
+        for sig in stop_val_refs:
+            if isinstance(ref, (ArrayReference, ArrayOfStructuresReference)):
+                raise GenerationError("{0} not yet supported in the stop "
```
In general, if it's a 'not yet supported' error then use `NotImplementedError`

arporter commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -562,6 +566,635 @@ def _encloses_omp_directive(self):
+        statements = schedule.walk((Assignment, IfBlock, Loop))
+        for statement in statements:
+            if isinstance(statement, Assignment):
+                lhs = statement.children[0]
```
Semantic nav: `.rhs` and `.lhs`.

arporter commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -562,6 +566,635 @@ def _encloses_omp_directive(self):
+                firstprivate_list.append(name)
+
+        # Look through the schedule and work out data sharing clauses.
+        statements = schedule.walk((Assignment, IfBlock, Loop))
```
Could this be implemented using a Visitor instead?

LonelyCat124 replied:
Yes I think so - but I'm not sure how to do this for PSyclone directives/where to put the visitor in the codebase. Is there an example I can look at?

arporter commented on src/psyclone/psyir/transformations/omp_task_trans.py:
```diff
@@ -0,0 +1,112 @@
+        :param options: a dictionary with options for transformations.
+        :type options: dict of string:values or None
+        '''
+        # Disallow CodeBlocks inside the region
```
If a Task can be applied to a region (and ,from my limited understanding, they can) then this transformation should inherit from RegionTrans which provides functionality for specifying those node types that are forbidden.

LonelyCat124 replied:
It now inherits from ParallelLoopTrans as for now I am only allowing them for Loops (at least for this transformation).

arporter commented:
I've had a quick look @LonelyCat124 and made some comments inline. I think you're suffering from the same problem that I have in #310 which is how we supply clauses to directives. Currently we are string based and have the `begin_string` method but this isn't very nice - I think (subject to a conversation with @sergisiso and @rupertford) that the content of clauses should be PSyIR expressions that appear as children of the directive node. Obviously that's not your problem right now but I did wonder whether you can get away with reducing your use of the FortranWriter?

LonelyCat124 commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -562,6 +566,635 @@ def _encloses_omp_directive(self):
+        # and create clause appropriately
+        index_name = "" 
+        if type(node.children[0]) is Reference:
+            index_name = writer(node.children[0].symbol)
```
Fixed this

LonelyCat124 commented:
> I've had a quick look @LonelyCat124 and made some comments inline. I think you're suffering from the same problem that I have in #310 which is how we supply clauses to directives. Currently we are string based and have the `begin_string` method but this isn't very nice - I think (subject to a conversation with @sergisiso and @rupertford) that the content of clauses should be PSyIR expressions that appear as children of the directive node. Obviously that's not your problem right now but I did wonder whether you can get away with reducing your use of the FortranWriter?

Thanks Andy for having a look through.
I think my difficulty with reducing use of FortranWriter is differentiating between `Reference` and `Literal` classes as Fortran (or language-independent) strings.

For simple `Reference` I can just do `ref.symbol.name`, but if its a `Literal` then I assume `literal.symbol.name` wouldn't work. Is there a "variable-independent" way to convert a Reference or Literal to its string representation?

LonelyCat124 commented:
Ok - now that clause is merged to master I'll look back into this next. I need to change OMPParallelDirective to use clauses and then OMPTaskDirective to do the same.

LonelyCat124 commented:
I started changing the `OMPParallelDirective` to use the new clause system.

I also made an `OMPReductionClause` but I realised I don't need this yet, but might not need it after all for this PR, as I don't think I'm supporting task reductions right now.

I'll review the failing tests tomorrow probably and then would you be able to have a look at the Parallel directive after that @arporter ? I'd start work on the task directive using the new clause system in the mean time.

LonelyCat124 commented:
Ok, I fixed up the tests (at least locally...). Most of the test changes are changing the expected code from:
`clause1, clause2, clause3` to `clause1 clause2 clause3` - this was to keep aligned with @sergisiso and I's decision on not having this comma be generated.

The rest of the OMPParallelDirective is now implemented using clauses. The OMPParallelDoDirective uses clauses for non-reduction operations now as well if anyone has time to look at those updates (either @arporter or @sergisiso I guess).

There are a few things I'm not totally happy with right now:
1. The children are not created at initialisation (even if they would be replaced later). Part of my reasoning for this is that the code inside can change, so the declaration of private variables could change, but probably there should be some initial child in the correct position.
2. There is some complication I forsee (though not yet implemented) where the parent class for `OMPParallelDoDirective` is an `OMPDoDirective`. These will have disparate clause children options - perhaps a `gen_clauses` function (or something similar) would be a good idea here.
3. The "updates" to the `private` clauses happen inside code gen - i'm not sure if this is a problem or not but it doesn't seem to be for now.

I think otherwise this change is fairly transparent - let me know if extra comments or anything would be good, or if anything else you feel is particularly bad from a design pov.

LonelyCat124 commented:
I realised this morning there is another "special" type of clause I need to implement, which have a different structure to a normal clause, and before I create a specific visitor only for that clause, i thought I'd check if there were any other clauses that had this structure (I think reduction clauses might too?)

So the structure for this clause is:
`CLAUSENAME( OPERAND: VAR[, VAR]*)`. The basic clause visitor just handles clauses as:
`node.clause_string + "(" + ",".join([ self._visit(child in children) ]) + ")", which can't handle the `OPERAND :` section.

I think what might be better is to specialise and "OperandClause" node, which also has a `_operand` member, and its own visitor.

Is this a structure used in OpenACC/elsewhere as well? @sergisiso @arporter

arporter commented:
Hi @LonelyCat124 , yes, OpenACC has at least the `reduction` clause which is of that form. It looks as though `worker`, `vector` and `gang` also have an optional, similar form.

rupertford commented:
Yes, there are similarities. Essentially the pattern I'm trying to support is `NAME( ARG, ..., NAME=ARG,...)`. At the moment we have a parent node for `NAME` and it has children nodes for each `ARG`. The question is how to capture the optional `NAME=`. In this case the obvious solution is to add another child e.g. `NamedNode`, that has a name and a single child. In your case you could have a child called `NamedList` or similar which had your optional name. However, we were concerned that we would affect existing tree traversal code if we did this as children in a list are then different classes. So instead the solution I am currently looking at is to add an additional property to the `NAME` node which captures which children are named. That would be even simpler for you as you could add an optional property called `named_list` (or `operand` if it is always an operand) which could be `None` or `OPERAND` and would determine the way the writer outputs.

LonelyCat124 commented:
> As Rupert hints at the end, I would not consider the operand to be a named children or a children at all, just a simple class attribute with the value of the operand (this being probably an enum of the available types). Similar to the operand in the various Operation classes. Then as you proposed, a specialized visitor for this kind of node.

Yes this is the implementation I have locally (not sure if I pushed as its messed up with equality check stuff) which I think is best here.

LonelyCat124 commented:
Started building up tests for this to try to work out for correctness. There's one thing for sure that are problematic:
-  ArrayReference handling when indices are `Reference + Literal` is wrong. I turned it into arrayslice but it should be something else.

If the Reference is the parent loop's variable, it should be `parent_loop_var + parent_loop_step` (or a multiple of parent_loop_step) if the parent_loop_step is a Literal. If the parent loop step is a Reference, then I won't support a binary operation being used for the variable for now (as I don't know what the step is).
If the Reference is the current loop variable, then it can be ignored for both BinaryOperation or just Reference, and should instead just be a `:` Range Node.

LonelyCat124 commented:
Ok, a lot of sanity things have been fixed to do with ArrayReferences.
`depend(in:` clauses now correctly use `:` instead of References when a loop variable of a child loop. I've not finished running the test yet, but there should be code to handle binary operations now as well.
`depend(out:` clauses do not yet use `:`. It should just be a matter of working the same strategy into the write reference handling code, but I'm not 100% on that.

This code works by checking `ref.symbol in child_loop_vars`, and if so then creates a Range object using LBOUND and UBOUND as appropriate.

LonelyCat124 commented:
Ok, I fixed that up. There is some extra difficulty now.

When I have a block of code, e.g.:
```
!$omp task
do i=1,10,1
do j = 1,10,1
  a(i,j) = k
  a(i,j) = b(i,j)
end do
end do
!$omp end task
```
Then my out dependency is being constructed wrong due to repeated entries, i.e. I get `depend(out: a(:,:),a(:,:))`

For indexes that are not ranges, this works fine as the check takes place and index == index2 is checked for Reference/Literal and seems ok (NB This is NOT maybe true for literal as that doesn't have an overriden `__eq__` function.

I need to work out how Ranges work for the `:` thing internally as I don't understand it so I can check them as well (slash implement __eq__)

LonelyCat124 commented:
Ok, looking back at this again today, and found a big oversight in my `handle_binary_op` strategy, which I'm unclear on the way to fix.

My assumption had been that if I had some code inside my task, e.g.:
```
do j=1, 320, 32
  !$omp task ....
  do i=j, j+32
    a(i) = b(i+1) - 3
  ...
```
was the clause(s) for the access to `b` could be handled by a single clause, however this is not the case. All clauses need to access a binary opreation of `j` and a multiple of the outer loop's step size. So in this case, to cover all values of `b` accessed by a single iteration, the correct clause needs to be `depend(in: b(j),b(j + 32))`. Right now I have no way to handle this, and it gets even more complex if there are more nested loops, e.g.

```
do jj=1, 320, 32
  do ii = 1, 320, 32
    !$omp task ...
    do i=ii, ii+32
      do j=jj, jj+32
       a(i, j) = b(i+1, j+1)
...
```
the required clauses is probably `depend(in: b(i,j), b(i+32, j) b(i, j+32), b(i+32, j+32))`.

The current `handle_binary_op` adds the new index (i.e. a `BinaryOperation` or `Reference`) to the `index_list`, which doesn't work when multiple are required. Instead I think the `index_list` needs to be a list of size N (where N is the number of indices) where each element is either:

- A `BinaryOperation`
- A `Reference` (or subclass)
- A `Literal`
- A `list`

The final set of index_lists can then be constructed from these. So if we had:
```
[ binop1, [binop2, binop3]]
```
the clauses would be: `(binop1, binop2)` and `(binop1, binop3)`, or for a worse case:
```
[ [binop1, binop2], [binop3, binop4] ]
```
the clauses would be `(binop1, binop3)`, `(binop2, binop3)`, `(binop1, binop4)`, `(binop2, binop4)`.

The difficulty (from a design perspective) I have right now is I don't have a good strategy to construct all the permutations efficiently but I will continue to think about it.

LonelyCat124 commented:
I fixed the previous issue, and have run into another oversight I need to fix. A normal use case might be something like:
```
do i = 1, 320
  do j = 1, 32
    a(i,j) = b(i+1, j) + k
  enddo
enddo
```
or similar.
A simple strategy in this case, would be to chunk the outer loop and create tasks along it:
```
do i = 1, 320, 32
  !$omp task firstprivate(ii), private(i, j), ....
  do ii = i, i + 32
    do j = 1, 32
      a(ii,j) = b(ii+1, j) + k
    enddo
  enddo
enddo
```

In this instance, `ii` is essentially a placeholder for the relevant value of `i`, so for the purposes of generating the dependencies it needs to be treated as such. I forgot about this when creating a lot of the code generation (oops) so its not taken into account  I don't think. I will have to go through the code and find out a good way to handle this.

LonelyCat124 commented:
Yet another issue I need to fix (I'm closing in on thinking of doing a full redesign for this class soon, the rate I'm thinking of new cases when writing tests doh, though I've learned a lot about what I might want to handle at least, and a better way of seperating things would probably be good).

Its probably also important/useful to potentially be able to support double chunked loops (does this seem useful/important to you @sergisiso ), e.g.
```
do i = 1, N, 32
  do j = 1, N, 32
  !$omp task ...
    do ii=i, i+32
      do jj=j, j+32
        A(ii, jj) = B(ii+1, jj+1) * 0.35215
...
```

This would currently work badly for a variety of reasons:

1. The parent loop variable is `j`, and since the initialisation of the node's first loop is not `loop_var = parent_loop_var` it can't tell that `ii` is a proxy for `i`, so will not correctly create the dependencies based on `i`.
2. Since the node's first loop var is not `jj`, it also can't tell that `jj` is a proxy for `j`, so will not correctly create the dependencies based on `j`


I think the dependencies computed in this case would be `depend(in: B(:,:))` and `depend(out: A(:,:))` which both are nonsensical.

I think I will write a bunch of tests to work out what code generation I want to be able to support and then come up with a design and go from there, just trying to fit "fixes" into the original code is getting messy.

LonelyCat124 commented:
I'm probably going to put design notes for the new ideas I have in this PR now.

Operation order:

1. Find all parent loops inside the same parallel region. Its not entirely clear how to do this efficiently, but probably it just has to involve recursive calls to `ancestor` using either Loop or OMPParallelRegion for the type argument. For each Loop, pull out the symbol and store it for later, and set it to be firstprivate.
2. Find all child loops inside the task region. For each of those loops, check if they are the inner section of a chunked loop structure, i.e. are they initialised as `do var = parent_loop_var, ...`. If so, then we can store that Loop's variable as a "proxy" for the relevant parent_loop_var (i.e. when encountering this variable, it should be treated as the parent loop's variable for the purposes of dependency analysis). The loop variable itself is declared as private. Notably this is not totally straightforward if there are multiple loops using the same loop variable, this "proxy-ing" should be done on a per-loop basis, and only apply to the children of that loop.

The requirement to handle 2 correctly makes the next step more complex than the initial implementation.

3. We want to find all children of the node's immediate Loop child (to which the transformation was applied) which are either: `Loop`, `Assignment`, `IfBlock`. 
4. For `Loop` nodes, we need to work out if they are a proxy loop, and then recurse through their children to find all `Assignment`/`IfBlock`/`Loop` nodes. 
5. For `Assignment` nodes, we perform basically the same operation as now, computing all writes to `shared` variables into out dependencies, and all reads to `shared` variables as in dependencies.
6. For `IfBlock` nodes we perform the same operations as now, computing the relevant in dependencies and private/firstprivate if appropriate.

I'm hopeful that these 6 steps are sufficient to handle the full functionality needed for tasking.

LonelyCat124 commented:
Ok, so the un-reimplemented functions so far are:
`handle_readonly_reference`
`handle_binary_op`

There's also all the handling of written-to-references and array accesses which was previously not in separate functions that now needs handling. I'm probably going to write the design of these in the PR this afternoon while I try to work out a good way to do them.

LonelyCat124 commented:
`handle_binary_op` at the moment does the following things, and is called only when a binary operation is found inside an array index:
1. If the binary operation is not an `ADD` or `SUB`, throws an exception
2. if the binary operation is not a `LITERAL OP REFERENCE` (or the opposite order) then throw an exception. This is possibly too strict (as one could do `LITERAL OP LITERAL` safely), but since those cases are unusual code I'm going to ignore them for now.
3. Find which of the children are the `Reference`, and store some info based on that for reconstructing the operation required for the clauses later.
4. There are then 2 options for how to handle this, and it depends on whether the `Reference` is to one of the proxy `parent_loop_var`s or not.

If the `Reference` is to a `parent_loop_var` proxy, we know that the `Reference` must be private (and thus in `private_list`).
In this case, we need to compute the set of clauses required, based upon the `Literal`'s value and the loop step size for the proxy variable.
Note: At the moment the variable has no way to reference the loop that it is the variable to, so we should store that Loop node in the proxy system too.

If the `Reference` is to a loop variable that is private and not a proxy, then we need to create a depend clause to that entire array (i.e. `:`) for that index.
If the `Reference` is to any other private variable, we assume its just a constant so just use a copy.

If the `Reference` variable is in `firstprivate_list`, then it currently checks if the variable is from a parent loop. If it is, then we use the same code as for when its a proxy variable, which I think is wrong. We can instead just use acopy the reference (since the parent loop variable is constant). I think this was done to handle the proxy variables from before.
If it is a child loop's variable, it should throw an exception (child loop variables are not allowed to be firstprivate). In fact I think this can never happen, so its mostly a sanity check for now, as its untestable code I think.
Otherwise, just use a copy of the reference.

Finally, if the variable has not yet been set private or firstprivate, we set it to firstprivate and then perform the same operations as for when its firstprivate.

If the variable is shared, then we throw an exception for now. If shared variables are used, they're in principle supportable, but I don't have a good use-case (since shared variables for now are ArrayReferences, which are banned from being used to index inside task directives).

I think I will split this function into functions to handle private and firstprivate accesses seperately.

LonelyCat124 commented:
`handle_read_only_ref` at the moment does the following things:
1. For `(ArrayReference, ArrayOfStructuresReference)`: At the moment, the initial implementation checks whether this ArrayReference was private in the parent parallel. This can never happen so can be removed. We thus know the Array is shared, so we loop over the indices. If an `Array, AoS or StructureReference` appears in the indexing we throw an exception. If the index is a `Reference`, handle computing the dependence & index structure accordingly, depending on if its a child loop, proxy loop or general variable. If its a binary operation then the `handle_binary_op` function is used. Anything else is disallowed (though I think a Literal should also be fine...). This results in a list of (lists) of indexes, and itertools.product is used to create all sets of clauses required to construct
2. For `StructureReference`: Create a reference to the symbol of the reference (for use later). Perform the same checks done for a general `Reference` on this new reference.
3. For `Reference`: If the variable was declared private in the parent parallel region, and is not yet in the private_list or firstprivate_list for this clause, add it to the firstprivate_list (as it is read before being written to, so we need some data). If it was not declared private in the parent parallel region, it is a shared variable, so we need to add an in dependency to this `Reference`.

LonelyCat124 commented:
Handling written to reference (i.e. LHS of an assignment) is currently in-line, and not in standalone functions, which I will change (so the code is easier to read and more modularised).

This is very similar to handling reads to variables, but in dependencies are instead out dependencies, and variables declared firstprivate are instead declared private. If there are other major differences I will comment them in here later.

LonelyCat124 commented:
Ok, the previous tests I wrote are fully working now with the new implementation.
One thing I forgot to test is the behaviour of if conditions, so those are next to write tests for. I'm fairly certain that functionality isn't finished yet either.

LonelyCat124 commented:
@arporter Hi Andy - if you have time to do another quick look over/review on the changes in omp_clauses.py and omp_directives.py (ignoring the old commented out code that I will remove later) at this stage it would be appreciated again!

In the mean time I will work on the transformation and tests.

LonelyCat124 commented:
TODO: Test for OMPParallelDirective to cover the addition of a new private clause in gen_code. Probably easiest to test by changing the tree after a gen_code for the first bit of if, and by deleting the private_clause and running again.

Same issue for begin_string - testing both might be difficult.

I'm not sure why this code isn't part of the "diff" on this patch

TODO: OMPParallelDirective._get_private_list is not covering variable names for child calls - I don't know why this is no longer covered but was before.

TODO: OMPTaskDirective._validate_child never checks the `False` part.
TODO: OMPTaskDirective doesn't check GenerationErrors occur for Binary Operation of types MUL/DIV/etc. or if there is Reference+Reference for example.
TODO: OMPTaskDirective doesn't check Generating code for Literal+Reference in BinaryOperation. Need to follow up after tthat is implemented.
TODO: OMPTASkDirective doesn't check GenerationError occurs for Binary Operation with a shared variable.
TODO: OMPTaskDirective read-only arrayreference doesn't have a test when the array reference is not yet declared firstprivate.
TODO: OMPTaskDirective read-only arrayreference doesn't have a test where the index is just some Reference.
TODO: OMPTaskDirective read-only arrayreference doesn't have a test where the index is shared (*difficult*).
TODO: OMPTaskDirective read-only reference doesn't have a StructureReference test.

TODO: OMPTaskDirective write ArrayRef doesn't test Literal index.
TODO: OMPTaskDirective write ArrayRef doesn't test shared index.

TODO: OMPTaskDirective write_baseref and write_reference for Structure/Reference not covered.
TODO: OMPTaskDirective error code in evaluate_loop not covered.
TODO: OMPTaskDirective not yet decalred firstprivate code in evaluate_loop not covered.
TODO: Else body not covered
TODO: compute_clauses error code not covered/written.

TODO: gen_code not covered (...)

arporter commented on src/psyclone/psyir/nodes/omp_clauses.py:
```diff
@@ -115,3 +117,223 @@ class OMPNogroupClause(Clause):
+    @property
+    def _clause_string(self):
+        clause_string = "default("
+        if self._clause_type == OMPDefaultClause.DefaultClauseTypes.SHARED:
```
I think there's some way of getting the name of a member of an enumeration so you don't have to do this.

LonelyCat124 replied:
I switched this to use name.lower()

arporter commented on src/psyclone/psyir/nodes/omp_clauses.py:
```diff
@@ -115,3 +117,223 @@ class OMPNogroupClause(Clause):
+    def set_schedule(self, schedule):
+        self._schedule = schedule
+
+    def __init__(self, schedule="static"):
```
Should we have an enumeration of valid schedule values too or is that complicated?

LonelyCat124 replied:
I was going to leave this for now since the existing code uses a string for the input, and there's no checking. I could add it but I think it should be a separate issue as this PR is already going to be large.

arporter commented on src/psyclone/psyir/nodes/omp_clauses.py:
```diff
@@ -115,3 +117,223 @@ class OMPNogroupClause(Clause):
+    def _validate_child(position, child):
+        '''
+         Decides whether a given child and position are valid for this node.
+         One child allowed, of type Literal.
```
That's not what's implemented :-)

LonelyCat124 replied:
Fixed docs oops!

arporter commented on src/psyclone/psyir/nodes/omp_clauses.py:
```diff
@@ -115,3 +117,223 @@ class OMPNogroupClause(Clause):
+
+    @property
+    def operand(self):
+        if self._operand == OMPDependClause.DependClauseTypes.IN:
```
Again, I think you can get the name of a member of an Enum and thus do away with this if block.

LonelyCat124 replied:
I switched this to use name.lower()

arporter commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -638,6 +705,946 @@ def _encloses_omp_directive(self):
+               self._evaluate_node(child_node, private_list, firstprivate_list,
+                                   shared_list, in_list, out_list)
+
+    def _evaluate_node(self, node, private_list, firstprivate_list,
```
Possibly, you could tidy this up by implementing a Visitor class that inherits from `psyclone.psyir.backend.PSyIRVisitor`. Your `_evaluate_assignment` would then be `assignment_node`, `loop_node` etc. etc. I say 'possibly' because I'm not totally certain that PSyIRVisitor will fully support what you're doing but it's definitely worth a look. For examples of implementations see e.g. `psyclone.psyir.backend.c` or `psyclone.psyad.adjoint_visitor`.

arporter commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -1028,13 +2063,26 @@ def gen_code(self, parent):
         zero_reduction_variables(calls, parent)
-        private_str = ",".join(self._get_private_list())
+        private_clause = self._get_private_list()
+        if len(self._children) >= 3 and private_clause != self._children[2]:
```
Is this just ensuring that the PSyIR is self-consistent at this point?

LonelyCat124 replied:
Yes, and make sure the private clause is in the right place/doesn't remove another clause etc.

arporter commented on src/psyclone/psyir/transformations/__init__.py:
```diff
@@ -99,4 +102,6 @@
            'HoistTrans',
            'OMPTaskwaitTrans',
-           'LoopTiling2DTrans']
+           'LoopTiling2DTrans',
```
I suspect (but haven't discussed with @sergisiso or @rupertford ) that we don't need to list the classes - this is only required if we do `from xxx import *` and we never do that - and it causes errors when building the reference guide.

arporter reviewed (changes requested):
I've only had a quick look through - I've not tried to understand the details of the implementation. I think it's looking a lot better now. The only significant thing left to investigate is whether having a sub-class of the `PSyIRVisitor` would help.

LonelyCat124 commented on src/psyclone/psyir/nodes/omp_clauses.py:
```diff
@@ -115,3 +117,223 @@ class OMPNogroupClause(Clause):
+        3. Same number of children.
+        4. Their children are equal.
+        '''
+        is_eq = super().__eq__(other)
```
I think in general we decided not, the follow statement doesn't run past `is_eq = is_eq` and it adds a bunch of code bloat. In this case I could probably reduce it to a single line, but it keeps the same structure as the other implementations this way.

LonelyCat124 commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -548,8 +606,9 @@ def _get_private_list(self):
 
-        :returns: list of variables to declare as thread private.
-        :rtype: list of str
+        :returns: A private clause containing the variables that need to be
```
Fixed this.

LonelyCat124 commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -638,6 +705,946 @@ def _encloses_omp_directive(self):
+        :rtype: bool
+
+        '''
+        if position == 0:
```
We could do any order probably (though it'd be a bit messy), but I didn't know a good way to create `_children_valid_format` for an "any order". The Directive also makes its children on creation to ensure ordering is correct for now.

LonelyCat124 commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -638,6 +705,946 @@ def _encloses_omp_directive(self):
+                "OMPTaskDirective must be inside an OMP Serial region "
+                "but could not find an ancestor node.")
+
+    def _find_parent_loop_vars(self):
```
I thought it might be potentially called quite frequently and could be quite heavy so I thought just call it once and store the solution, but I think we can make a final decision on that later.

LonelyCat124 commented:
Have issue with ArrayReferences not being equal in some cases, need to investigate further.

LonelyCat124 commented:
Fixed the bug from yesterday, and improved coverage more. I need to still handle:

- [ ] private variable indexes - no test code exists yet to handle this because i've not quite worked out making private variables reliably yet, but maybe I have some code that does this elsewhere in the tests for other things.
- [ ] StructureReference read access
- [ ] StructureReference write access
- [ ] Literal value in array write indexing
- [ ] Shared loop variable + code to do error message
- [ ] Array reference in loop start val, loop stop val, loop step val
- [ ] Step val not yet declared firstprivate
- [ ] Directive with muiltiple children + error message
- [ ] Directive with non-loop child + error message.

There is currently a `gen_code` implementation - @arporter/@sergisiso would you recommend just deleting this or can I leave it in but without code coverage?

arporter commented:
> There is currently a gen_code implementation - @arporter / @sergisiso would you recommend just deleting this or can I leave it in but without code coverage?

I'm increasingly of the opinion that un-tested code is  (or very soon will be) broken code so I would vote for removing it.

LonelyCat124 commented:
Almost successfully have covered things now.

Remaining:

- [ ] Non-proxy access to loop variable in a binary operation where the divisor is > 2
- [ ] Non-proxy access to loop variable in a binary operation where modulo is not 0 and divisor > 2
- [ ] Declared firstprivate variable accessed in array reference - difficult code path to get PSyclone to create.
- [ ] Undeclared firstprivate variable accessed in array reference - same as above but not yet accessed.
- [ ] Shared (non-arrayref) variable accessed in array reference.
- [ ] Shared (non-arrayref) variable used as a loop variable - also needs error message.

I'll probably come back to this in a few weeks though as I need to work on other projects again.

Once the coverage of the directive is done and I'm happy with the implementation, the remaining things to do in this PR:

- [ ] Delete the commented-out ParalleLoopTrans from transformations.py
- [ ] Code cleanup (pycodestyle/pylint is significantly behind here)
- [ ] Implementation of dependency analysis/checking in OMPSingleRegion.

LonelyCat124 commented:
I'm hoping to have a look at this again in the upcoming week or two, hopefully I can get onto the dependency analysis/checking and have this ready for review.

LonelyCat124 commented:
@arporter @sergisiso @rupertford I'm not sure whether I should put this ready to review soon or wait.

The main remaining piece of work to do in terms of new implementation is "Implementation of dependency analysis/checking in OMPSingleRegion."

Fundamentally, the transformation and directive in this PR now work correctly (at least, I believe they do). They could be used and would generate correct fortran code if provided good input. 
The remaining piece is to ensure that PSyclone throws an Error instead of generating code that doesn't preserve the serial output when transformations are applied that will not preserve correctness under OpenMP's dependency structure.

I felt previously that this PR should therefore include that, as otherwise it would be possible to use PSyclone to create invalid code (which PSyclone i think is not meant to do).

However, this PR already includes so much "stuff" that I wondered if asking someone to review this + that would be too big, and the effect on other ongoing work might cause too many conflicts so I am wondering whether this should be done, and I start work on the other code while this is being reviewed (which I expect may take some time anyway). This could be approved and either merged or just kept up-to-date with master independently of the other work if needed.
Right now this PR includes:
1. Implementation of OMPTaskDirective (1000 loc still after refactoring/restructuring) & OMPTaskTransformation + the 30 or so tests of validity.
2. Removal of ParallelLoopTrans from transformations.py to its own file in `psyir/transformations`
3. 7 new OpenMP clause classes
4. Updating code clause generation to not place unnecessary `,` between OpenMP clauses. The structure of directive in OpenMP is now `!$omp parallel [clause[ [,] clause] ... ] ` and the OpenMP examples I've seen provided by OpenMP do not use `,` between clauses. I can revert this if preferred though, I'm not entirely sure why I ended up changing it any more.

What do you think the best approach is here?

arporter commented:
I'm all in favour of trying to keep PRs small. Although we do aim to prevent PSyclone generating incorrect code, I think in this instance a warning or two in the user guide would be sufficient (with a reference to where the dep. analysis is being tackled).
In this vein, you could also perhaps break out the commas issue into a separate ticket (if you feel strongly about it). We possibly all need to agree what format we're happy with as I suspect we're in danger of flip-flopping between different options :-)

LonelyCat124 commented:
I will look again and try to work out if there was a specific reason I wanted to remove the commas from directives today. I don't think there should be a specific reason to do it and I'd be surprised if either OpenMP or OpenACC decide to stop supporting the commas.

I'll try to fix up the bulk of this code while i'm travelling tomorrow so its closer to review as well.

LonelyCat124 commented:
I reverted the comma removal, while fixing coverage I did spot another "issue" with reductions still not using clauses, and I've made a separate issue for that independently of this (#1812) as I'm not worrying about reductions in task code for now.

LonelyCat124 commented:
Ok, everything fixed up in this in terms of functionality now. I'll work on docs and code cleanup and then mark it ready for review.

LonelyCat124 commented:
Ok, I think this is now ready for a first real review.
Some of the previous comments may still apply, but I think the code has changed enough from the initial implementation that many of them may not be relevant any more.

LonelyCat124 commented:
One thought I have had (which can wait a bit) is that I may benchmark the code as it is now, or with some manual changes to change the dependency clauses that are currently generated as `A(i)` to `A(i:i+X)` and see if there is much performance difference. Functionally the way the code is written there should be no change, but it is more obvious what is happening when reading the generated code.

I erred from it initially as I have enountered performance issues in the past.

LonelyCat124 commented:
Forgot to tag @arporter @sergisiso about review - not sure which of you feels better positioned to take a look.

sergisiso commented on src/psyclone/tests/psyir/transformations/transformations_test.py:
```diff
@@ -53,7 +53,7 @@
 from psyclone.psyir.tools import DependencyTools
 from psyclone.psyir.transformations import ProfileTrans, RegionTrans, \
-    TransformationError
+    TransformationError, ParallelLoopTrans
```
pylint says ParallelLoopTrans is not used

LonelyCat124 replied:
Fixed.

sergisiso commented on src/psyclone/tests/psyir/transformations/transformations_test.py:
```diff
@@ -533,6 +533,7 @@ def test_omplooptrans_apply(sample_psyir, fortran_writer):
   !$omp end loop
   !$omp end parallel\n'''
 
+    print(fortran_writer(tree))
```
Remove print

LonelyCat124 replied:
Fixed.

sergisiso reviewed (changes requested):
@LonelyCat124 See the first set of clean up suggestions and I also suggest to move some parts of it into a separate PR if it is easy to do (more details inline).

LonelyCat124 commented on src/psyclone/psyir/nodes/omp_clauses.py:
```diff
@@ -115,3 +117,222 @@ class OMPNogroupClause(Clause):
+
+class OMPReductionClause(OperandClause):
+    '''
+    OpenMP Reduction clause. Not yet used
+    '''
+    _children_valid_format = "[Reference]+"
+    _text_name = "OMPReductionClause"
+    # FIXME Reduction string and operator 
```
This class now just says `# TODO: #1812 Reduction string and operator`. I introduced this Clause as reductions should move into a clause in the future (and tasking may need to support reductions). I don't really understand how the reductions are generated for now so I have avoided implementing the details here or using the class outside of tests, but I wanted it in place for validating children correctly in classes which use reduction clauses.

Does that seem reasonable?

LonelyCat124 commented on src/psyclone/psyir/backend/c.py:
```diff
@@ -448,6 +448,29 @@ def loop_node(self, node):
+        :param node: a Clause PSyIR node.
+        :type node: :py:class:`psyclone.psyir.nodes.Clause`
+
+        :returns: the Fortran code for this node.
```
fixed

LonelyCat124 commented on src/psyclone/psyir/backend/c.py:
```diff
@@ -461,7 +484,22 @@ def regiondirective_node(self, node):
+            val = self._visit(clause)
+            # Some clauses return empty strings if they should not
+            # generate any output (e.g. private clause with no children).
+            if not (val.isspace() or val == ""):
```
The isspace can be removed it seems so I've changed that and all tests still pass.

LonelyCat124 commented on src/psyclone/psyir/backend/fortran.py:
```diff
@@ -1537,6 +1537,34 @@ def nemokern_node(self, node):
+    def operandclause_node(self, node):
+        '''This method is called when a OperandClause is
+        found in the PSyIR tree. It returns the clause and its children
+        as a string
```
Fixed.

LonelyCat124 commented on src/psyclone/psyir/backend/c.py:
```diff
@@ -448,6 +448,29 @@ def loop_node(self, node):
                f"{variable_name}<={stop}; {variable_name}+={step})\n"\
                f"{self._nindent}{{\n{body}{self._nindent}}}\n"
 
+    def clause_node(self, node):
```
I think they are, I can move this there in that case.

LonelyCat124 replied:
Fixed.

LonelyCat124 commented on src/psyclone/psyir/nodes/clause.py:
```diff
@@ -57,3 +58,19 @@ def clause_string(self):
+
+class OperandClause(Clause, metaclass=abc.ABCMeta):
+    '''
+    Base abstract class for all clauses that have an operand
```
Fixed

LonelyCat124 commented on src/psyclone/psyir/nodes/omp_clauses.py:
```diff
@@ -37,8 +37,10 @@
 ''' This module contains the implementations of the various OpenMP Directive
```
Fixed.

LonelyCat124 commented on src/psyclone/psyir/nodes/omp_clauses.py:
```diff
@@ -115,3 +117,222 @@ class OMPNogroupClause(Clause):
+    _text_name = "SharedClause"
+
+    @property
+    def _clause_string(self):
```
Discussed in person and decided it doesn't matter either way, my preference is this implementation so I will not change it.

LonelyCat124 commented on src/psyclone/psyir/nodes/omp_clauses.py:
```diff
@@ -115,3 +117,222 @@ class OMPNogroupClause(Clause):
+class OMPSharedClause(Clause):
+    '''
+    OpenMP shared clause. This is used to declare variables as shared in an
+    OpenMP region
```
I think I got it all now

LonelyCat124 commented on src/psyclone/psyir/nodes/omp_clauses.py:
```diff
@@ -115,3 +117,222 @@ class OMPNogroupClause(Clause):
+class OMPDefaultClause(Clause):
+    '''
+    OpenMP Default clause. Used to determine the default declaration for
+    variables used in an OpenMP region.
```
Fixed.

LonelyCat124 commented on src/psyclone/psyir/nodes/omp_clauses.py:
```diff
@@ -115,3 +117,222 @@ class OMPNogroupClause(Clause):
+    _children_valid_format = None
+    _text_name = "OMPDefaultClause"
+
+    @property
```
I think __init__ is the first method in all classes in here now.

LonelyCat124 commented on src/psyclone/psyir/nodes/omp_clauses.py:
```diff
@@ -115,3 +117,222 @@ class OMPNogroupClause(Clause):
+        return clause_string
+
+
+    def __init__(self, clause_type=DefaultClauseTypes.SHARED):
```
Fixed (I assume I don't need to document **kwargs).

LonelyCat124 commented on src/psyclone/psyir/nodes/omp_clauses.py:
```diff
@@ -115,3 +117,222 @@ class OMPNogroupClause(Clause):
+        INOUT=2
+
+    def __init__(self, depend_type=DependClauseTypes.INOUT):
+        if not isinstance(depend_type, OMPDependClause.DependClauseTypes):
+            raise TypeError(
+                    "OMPDependClause expected 'depend_type' argument of type "
+                    "OMPDependClause.DependClauseTypes but found '{0}'"
+                    .format(type(depend_type).__name__))
+        self._operand = depend_type
```
I don't think so, I don't see a reason for the `operand` property to be modifiable after instantiation in the Depend clause so keeping it private is better I think.

LonelyCat124 commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -435,23 +444,60 @@ def end_string(self):
+    _children_valid_format = ("Schedule, OMPDefaultClause, [OMPPrivateClause], "
+                              "[OMPReductionClause]*")
+
+    def __init__(self, children=None, parent=None):
```
Fixed.

LonelyCat124 commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -435,23 +444,60 @@ def end_string(self):
+                              "[OMPReductionClause]*")
+
+    def __init__(self, children=None, parent=None):
+       super(OMPParallelDirective, self).__init__(children=children, parent=parent)
```
@sergisiso Does this mean `super(OMPParallelDirective, self).__init__(**kwargs)` just become `super()` or is it `super().__init__(**kwargs)`?

LonelyCat124 replied:
No longer relevant

LonelyCat124 commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -435,23 +444,60 @@ def end_string(self):
+
+    def __init__(self, children=None, parent=None):
+       super(OMPParallelDirective, self).__init__(children=children, parent=parent)
+       self.addchild(OMPDefaultClause(clause_type=
```
This is to mirror the current behaviour with clause generation. At the moment the `OMPParallelDirective` will always generate `!$omp parallel default(shared)`, so I add the child in the constructor to replicate that behaviour.

LonelyCat124 replied:
The create method works. The one slight bit of weirdness is in `OMPParallelTrans` there is now:

```
         # Set the type of directive that the base class will use
         self._pdirective = OMPParallelDirective.create  
```

since we need to use the `create` method instead of the init function, but as long as you're ok with that it should work.

LonelyCat124 replied:
Do I need to add any docs to say to use the `create` method instead of the constructor?

sergisiso replied:
I see there are some inits and some create in use now
```
src/psyclone/transformations.py
1473:        self._pdirective = None
1606:        self._pdirective = self._directive
1738:        self._pdirective = OMPMasterDirective
1793:        self._pdirective = OMPParallelDirective.create
1866:        self._pdirective = ACCParallelDirective
```

but this ended up looking as an OOP factory pattern I think. Maybe renaming the attribute to `self._directive_factory` will make the intent more clear and already imply that any method that returns a new directive can be provided. What do you think?

Also you will need to tweak the comment preceding this assignment.

LonelyCat124 replied:
I think thats a better naming scheme. I've made this change and have updated the comment above I think.

LonelyCat124 commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -45,19 +45,28 @@
 from __future__ import absolute_import
 import abc
 import six
+import math
```
I think this is fixed.

LonelyCat124 commented on src/psyclone/transformations.py:
```diff
@@ -163,167 +168,164 @@ def validate(kern, options=None):
-
-        # Add the loop directive as a child of the node's parent
-        node_parent.addchild(directive, index=node_position)
+#class ParallelLoopTrans(LoopTrans, metaclass=abc.ABCMeta):
```
Done.

LonelyCat124 commented on src/psyclone/psyir/transformations/parallel_loop_trans.py:
```diff
@@ -0,0 +1,211 @@
+    # being transformed.
+    excluded_node_types = (nodes.Return, psyGen.HaloExchange, nodes.CodeBlock)
+
+    @abc.abstractmethod
+    def __str__(self):
+        return  # pragma: no cover
```
Done

LonelyCat124 commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -435,23 +444,60 @@ def end_string(self):
 class OMPParallelDirective(OMPRegionDirective):
     ''' Class representing an OpenMP Parallel directive. '''
 
+    _children_valid_format = ("Schedule, OMPDefaultClause, [OMPPrivateClause], "
+                              "[OMPReductionClause]*")
```
Discussed, need to change this so that `OMPPrivateClause` is non-optional child and always child 2 for now.

@sergisiso will discuss the structure of directive's in PSyclone developers' meeting to determine children order for these (i.e. `Schedule` first or clauses first).

sergisiso replied:
I am still not sure what is the nicest solution here. Another option we talked about in to have a ClausesList which is mandatory and always proceed the directive body. And inside having the clauses/ or nothing if all are optional.

But since I am still unconvinced and you need this to continue with your other PR. I am happy with the current implementation and I will come back to this in a future issue/pr.

LonelyCat124 replied:
I am still not sure what is the nicest solution here. Another option we talked about in to have a ClausesList which is mandatory and always proceed the directive body. And inside having the clauses/ or nothing if all are optional.

LonelyCat124 commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -435,23 +444,60 @@ def end_string(self):
             parent.add(DeclGen(parent, datatype="integer",
                                entity_decls=[thread_idx]))
-        private_str = ",".join(private_list)
+        if len(self._children) >= 3:
+            if private_clause != self._children[2]:
+                if isinstance(self._children[2], OMPPrivateClause):
+                    self._children[2] = private_clause
+                else:
+                    self.addchild(private_clause, index=2)
+        else:
+            self.addchild(private_clause, index=2)
```
I changed this according to the discussion we had, and moved the setting in gen_code to after the validation.

LonelyCat124 commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -543,14 +600,15 @@ def end_string(self):
         return "omp end parallel"
 
-    def _get_private_list(self):
+    def _get_private_clause(self):
```
Unchanged as discussed.

I also made `private_clause` and `default_clause` properties but I did not add setters as I feel like they probably shouldn't be publically modifiable.

LonelyCat124 commented on src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:
```diff
@@ -6303,7 +6311,7 @@ def test_intergrid_omp_parado(dist_mem, tmpdir):
     gen = str(psy.gen)
     assert "loop4_stop = ncolour_fld_c" in gen
-
+    print(gen)
```
Removed.

LonelyCat124 commented on src/psyclone/psyir/transformations/parallel_loop_trans.py:
```diff
@@ -0,0 +1,211 @@
+from psyclone.psyir.transformations.transformation_error import \
+    TransformationError
+
+class ParallelLoopTrans(LoopTrans, metaclass=abc.ABCMeta):
```
All linting issues should be done.

LonelyCat124 commented on src/psyclone/psyir/nodes/omp_clauses.py:
```diff
@@ -115,3 +117,222 @@ class OMPNogroupClause(Clause):
+                    "OMPDefaultClause.DefaultClauseTypes but found '{0}'"
+                    .format(type(clause_type).__name__))
+        self._clause_type = clause_type
+        super(OMPDefaultClause, self).__init__()
```
I think I added all of these.

LonelyCat124 commented on src/psyclone/tests/psyir/backend/fortran_test.py:
```diff
@@ -2304,7 +2304,7 @@ def test_fw_directive_with_clause(fortran_reader, fortran_writer):
     assert '''!$omp parallel default(shared), private(i)
   !$omp master
-  !$omp taskloop num_tasks(32) nogroup
+  !$omp taskloop num_tasks(32), nogroup
```
I don't - clause node is now covered by the same tests I guess (since its in language_writer).

I added a test for directive with clause in c now as well.

LonelyCat124 commented:
Ok, I think I fixed up most of the review comments and remerged with master.

I will extract the Task classes into new files and a new PR on Monday and re-refer both for review (though the task one will presumably show the changes from both PRs until this one is merged).

LonelyCat124 commented:
Looks like I removed some tests or some things aren't covered, I'll go through that and fix it up.

sergisiso commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -511,6 +585,14 @@ def gen_code(self, parent):
         self.gen_post_region_code(parent)
 
+    def lower_to_language_level(self):
+        ''' In-place construction of clauses as PSyIR constructs.'''
```
Can you extend this description to explain that in the higher level some directives relay on dynamic variable dependence logic to decide what is private/shared and then we use this lowering step to find out this references and place them explicitly to the lower level tree the the backend visitor will process.

LonelyCat124 replied:
Done.

sergisiso commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -543,14 +621,15 @@ def end_string(self):
 
-        :returns: list of variables to declare as thread private.
-        :rtype: list of str
+        :returns: A private clause containing the variables that need to be
```
backslash for line continuation

LonelyCat124 replied:
Added

sergisiso commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -543,14 +621,15 @@ def end_string(self):
         return "omp end parallel"
 
-    def _get_private_list(self):
+    def _get_private_clause(self):
```
Does this method assume a certain sharing default? e.g. default(shared) ? In this case, this PR introduced the OMPDefaultClause which can interfere with this and make this logic wrong if it is not the expected value?

Maybe a TODO in the docstrign and check/raise error is enough for now until we can think this better.

LonelyCat124 replied:
I'm not sure - the directive speficially adds a `OMPDefaultClause(clause_type=OMPDefaultClause.DefaultClauseTypes.SHARED))` to match the previous behaviour so I don't think any default is going to come up. I can add it though if you still think its needed.

sergisiso replied:
But the children can be changed. So:
```
dir = OMPParallelDirective.create(children)
dir.walk(OMPDefaultClause).replace_with(
    OMPDefaultClause(clause_type=OMPDefaultClause.DefaultClauseTypes.PRIVATE)
)
FortranWriter(dir)
```
Could this be wrong? Because it calls lower_to_language_level which calls _get_private_clause which adds a private clause, but in this case this is redundant and are the SHARED the ones that need to be explicitly listed.

LonelyCat124 replied:
It would be yeah - I've added a check for this now in `_get_private_clause` which has to be called for codegen to happen which should handle this.

sergisiso reviewed (changes requested):
@LonelyCat124 Thanks for splitting the PR. It makes the review much more manageable. I think this part should be almost there.

The most important remaining issue is considering if the clauses that are dynamically created in the lowering are assuming a certain OpenMP default and would be incorrect if we change it.

Also note that I replied in a couple of comments from my previous review, and that the branch need to be brought to master.

LonelyCat124 commented on src/psyclone/f2pygen.py:
```diff
@@ -141,7 +141,7 @@ class OMPDirective(Directive):
     def __init__(self, root, line, position, dir_type):
         self._types = ["parallel do", "parallel", "do", "master", "single",
-                       "taskloop", "taskwait", "declare"]
+                       "taskloop", "taskwait", "task", "declare"]
```
Fixed.

LonelyCat124 commented on src/psyclone/psyir/backend/c.py:
```diff
@@ -461,7 +461,21 @@ def regiondirective_node(self, node):
+        if len(clause_list) > 0:
+            result_list.append(" ")
+        result_list.append(", ".join(clause_list))
+        result_list.append("\n{\n")
```
Fixed.

LonelyCat124 commented on src/psyclone/tests/psyir/backend/fortran_test.py:
```diff
@@ -2317,3 +2317,15 @@ def test_fw_clause(fortran_writer):
+
+
+def test_fw_operand_clause(fortran_writer):
+    '''Test that a PSyIR operand clause is translate to the correct Fortran
```
Fixed

LonelyCat124 commented on src/psyclone/psyir/nodes/clause.py:
```diff
@@ -46,6 +46,7 @@ class Clause(Node, metaclass=abc.ABCMeta):
     '''
     _children_valid_format = None
     _text_name = "Clause"
```
Removed

LonelyCat124 commented on src/psyclone/psyir/nodes/omp_clauses.py:
```diff
@@ -115,3 +117,300 @@ class OMPNogroupClause(Clause):
+    OpenMP region.
+    '''
+    _children_valid_format = "[Reference]*"
+    _text_name = "SharedClause"
```
Removed

LonelyCat124 commented on src/psyclone/psyir/nodes/omp_clauses.py:
```diff
@@ -115,3 +117,300 @@ class OMPNogroupClause(Clause):
+    _text_name = "SharedClause"
+
+    @property
+    def _clause_string(self):
```
Added

LonelyCat124 commented on src/psyclone/psyir/nodes/omp_clauses.py:
```diff
@@ -115,3 +117,300 @@ class OMPNogroupClause(Clause):
+    OpenMP shared clause. This is used to declare variables as shared in an
+    OpenMP region.
+    '''
+    _children_valid_format = "[Reference]*"
```
Yeah sounds reasonable, I'll remove them.

LonelyCat124 commented on src/psyclone/psyir/nodes/omp_clauses.py:
```diff
@@ -115,3 +117,300 @@ class OMPNogroupClause(Clause):
+    OpenMP private clause. This is used to declare variables as private
+    to an OpenMP region.
+    '''
+    _children_valid_format = "[Reference]*"
```
Done

LonelyCat124 commented on src/psyclone/psyir/nodes/omp_clauses.py:
```diff
@@ -115,3 +117,300 @@ class OMPNogroupClause(Clause):
+    _text_name = "PrivateClause"
+
+    @property
+    def _clause_string(self):
```
Added

LonelyCat124 commented on src/psyclone/psyir/nodes/omp_clauses.py:
```diff
@@ -115,3 +117,300 @@ class OMPNogroupClause(Clause):
+    '''
+    OpenMP Schedule clause used for OMP Do Directives.
+
+    :param str schedule: The OpenMP schedule to use with this directive.
```
Added I think

LonelyCat124 commented on src/psyclone/psyir/nodes/omp_clauses.py:
```diff
@@ -115,3 +117,300 @@ class OMPNogroupClause(Clause):
+        '''
+        Gets the schedule of this OMPScheduleClause.
+
+        :returns: The schedule for this OMPScheduleClause.
```
Fixed

LonelyCat124 commented on src/psyclone/psyir/nodes/omp_clauses.py:
```diff
@@ -115,3 +117,300 @@ class OMPNogroupClause(Clause):
+        Set the schedule for this OMPScheduleClause.
+
+        :param str schedule: The schedule to use for this clause.
+        '''
```
Added

LonelyCat124 commented on src/psyclone/psyir/nodes/omp_clauses.py:
```diff
@@ -115,3 +117,300 @@ class OMPNogroupClause(Clause):
+    _schedule = ""
+
+    def __init__(self, schedule="static", **kwargs):
+        self._schedule = schedule
```
Fixed.

LonelyCat124 commented on src/psyclone/psyir/nodes/omp_clauses.py:
```diff
@@ -115,3 +117,300 @@ class OMPNogroupClause(Clause):
+    '''
+    OpenMP Depend clause used for OpenMP Task directives.
+
+    :param depend_type: The dependency type to use for this clause. The default
```
Done.

LonelyCat124 commented on src/psyclone/psyir/nodes/omp_clauses.py:
```diff
@@ -115,3 +117,300 @@ class OMPNogroupClause(Clause):
+    :param kwargs: additional keyword arguments provided to the PSyIR node.
+    :type kwargs: unwrapped dict.
+
+    :raises TypeError: If the supplied depend_type argument is the wrong type.
```
Fixed

LonelyCat124 commented on src/psyclone/psyir/nodes/omp_clauses.py:
```diff
@@ -115,3 +117,300 @@ class OMPNogroupClause(Clause):
+    OpenMP Default clause. Used to determine the default declaration for
+    variables used in an OpenMP region.
+
+    :param clause_type: The default data-sharing attribute to be described
```
Done

LonelyCat124 commented on src/psyclone/psyir/nodes/omp_clauses.py:
```diff
@@ -115,3 +117,300 @@ class OMPNogroupClause(Clause):
+        return isinstance(child, Reference)
+
+    @property
+    def operand(self):
```
Added

LonelyCat124 commented:
@sergisiso I think I went through the review and fixed the remaining issues. I also added the warning you recommended about the Clause class in the docs. Should be ready to review again.

sergisiso reviewed (approved):
@LonelyCat124 All the comments have been addressed. There are still some rough edges with Clauses but this comes from the fact that we still support gen_code and that LFRic dynamically generates its private variables list and needed to be accounted for the new classes. Hopefully this will get better if we modernize LFRic. Also there are some questions about the best ordering and optionality of clauses in the children list but this can be revisited in following PRs.

I made some small formatting and pylint clean ups. (Since I updated python and pycodestyle it started flagging new issues but I will address this in a separate PR).
